// @feiye/pinyin/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pinyinKeywords = exports.pinyinInitial = exports.pinyin = exports.pinyinObj = void 0;
const data_1 = require("./data");
/**
 * 汉子转换为拼音对象
 */
function pinyinObj(str, opt) {
    if (!str)
        return [];
    opt = Object.assign({
        pinyin: true,
        word: true,
        number: true,
        wordnumber: true,
        other: true,
    }, opt);
    const pinyin = [];
    let word = '';
    let number = '';
    let wordnumber = '';
    for (let i = 0; i <= str.length; i++) {
        let result;
        let type = 'other';
        const c = str.charCodeAt(i);
        const isNum = c >= 48 && c <= 57; // 48-57[0-9]
        const isAlp = (c >= 65 && c <= 90) || c >= 97 && c <= 122; // 65-90[A-Z] 97-122[a-z]
        if (isNum || isAlp) {
            if (opt.number && isNum) {
                number += str[i];
            }
            else if (opt.word && isAlp) {
                word += str[i];
            }
            if (opt.wordnumber) {
                wordnumber += str[i];
            }
            continue;
        }
        if (word && number) {
            result = wordnumber;
            type = 'wordnumber';
        }
        else if (word) {
            result = word;
            type = 'word';
        }
        else if (number) {
            result = number;
            type = 'number';
        }
        if (result) {
            pinyin.push({
                result,
                type,
            });
            result = '';
        }
        word = '';
        number = '';
        wordnumber = '';
        if (str[i] in data_1.default) {
            type = 'pinyin';
            if (opt.pinyin) {
                result = data_1.default[str[i]];
            }
            else {
                result = str[i];
            }
        }
        else if (opt.other && !result) {
            result = str[i];
            type = 'other';
        }
        result && pinyin.push({
            result,
            type,
        });
        if (opt.limit && pinyin.length >= opt.limit) {
            break;
        }
    }
    return pinyin;
}
exports.pinyinObj = pinyinObj;
/**
 * 字符串转换为拼音字符串
 * @param str
 * @returns
 */
function pinyin(str, opt) {
    return pinyinObj(str, opt).filter(i => i.type !== 'other').map(i => i.result).join(' ');
}
exports.pinyin = pinyin;
/**
 * 取得首字母
 * 注意：如果参数为字符串也只会返回首字母
 * @param str 字符
 */
function pinyinInitial(str) {
    const res = pinyinObj(str, {
        other: false,
        limit: 1,
    });
    if (res.length) {
        return res[0].result[0].toUpperCase();
    }
    return null;
}
exports.pinyinInitial = pinyinInitial;
/**
 * 字符串提取拼音关键词
 * @param str
 * @returns 结果是去重的
 */
function pinyinKeywords(str, opt) {
    var _a;
    const res = pinyinObj(str, opt);
    const out = new Set();
    for (let p = 0; p < res.length; p++) {
        const type = res[p].type;
        // 拼接前后拼音为词
        if (type === 'pinyin' && ((_a = res[p + 1]) === null || _a === void 0 ? void 0 : _a.type) === 'pinyin') {
            out.add(res[p].result + res[p + 1].result);
        }
        else if (type === 'word' || type === 'number' || type === 'wordnumber') {
            out.add(res[p].result);
        }
    }
    return Array.from(out);
}
exports.pinyinKeywords = pinyinKeywords;
// console.log(pinyinKeywords("中华人民共和国abcde0123，hello电梯 维修工，从事T22维修20年h3c world", { pinyin: false }))
